pkg/serviceaccounttoken/secret.go

```diff
@@ -153,14 +194,42 @@ func ServiceAccountSecret(ctx context.Context, sa *v1.ServiceAccount, secretList
 			logrus.Errorf("unable to delete secret [%s:%s]: %v", s.Namespace, s.Name, err)
 		}
 	}
+
 	return result, nil
 }
 
-func isSecretForServiceAccount(secret *v1.Secret, sa *v1.ServiceAccount) bool {
-	if secret.Type != v1.SecretTypeServiceAccountToken {
+func isSecretForServiceAccount(secret *corev1.Secret, sa *corev1.ServiceAccount) bool {
+	if secret.Type != corev1.SecretTypeServiceAccountToken {
 		return false
 	}
 	annotations := secret.Annotations
 	annotation := annotations[serviceAccountSecretAnnotation]
+
 	return sa.Name == annotation
 }
+
+func createServiceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface, lockKey string) (*corev1.Secret, error) {
+	if lockKey != "" {
+		mutex := getLock(lockKey)
+		mutex.Lock()
+		defer func(key string) {
+			mutex.Unlock()
+			lockMap.Delete(lockKey)
+		}(lockKey)
+	}
+
+	// We could have been waiting for the Mutex to unlock in a parallel run of
+	// createServiceAccountSecret - check again for the secret existing.
+	secret, err := serviceAccountSecret(ctx, sa, secretLister, secretClient)
```

Comment:
We don't seem to be checking this error - any way that we can check for an error and return early? Notably, I'm thinking that this could produce some extra secrets if the list call fails and we create a secret anyway (though this is unlikely in production).